crypto/BRCryptoWalletManagerClient.c

```diff
@@ -1286,19 +1319,25 @@ cwmTransactionEventAsETH (BREthereumClientContext context,
                                                          { .state = { oldState, newState }}
                                                      });
 
-                cryptoTransferSetState (transfer, newState);
                 cwm->listener.transferEventCallback (cwm->listener.context,
                                                      cryptoWalletManagerTake (cwm),
-                                                     wallet,
-                                                     transfer,
+                                                     cryptoWalletTake (wallet),
+                                                     cryptoTransferTake (transfer),
                                                      (BRCryptoTransferEvent) {
                                                          CRYPTO_TRANSFER_EVENT_DELETED
                                                      });
             }
             break;
+        }
     }
 
-    // TODO: crypto{Wallet,Transfer}Give()
+    if (NULL != transfer) {
+        cryptoTransferGive (transfer);
+    }
+
+    if (NULL != wallet) {
+        cryptoWalletGive (wallet);
+    }
```

Comment:
Okay.  I see this whole function as being explicit, clearer in argument passing.   (In the switch cases one doesn't have to track "the last call shouldn't take" - just to avoid this final give.)